src/docker/remote.rs

```diff
@@ -0,0 +1,653 @@
+        if file.file_type()?.is_file() {
+            fs::copy(&src_path, &dst_path)?;
+        } else {
+            fs::create_dir(&dst_path).ok();
+            copy_dir(&src_path, &dst_path, depth + 1, skip)?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn copy_volume_container_rust(
+    engine: &Engine,
+    container: &str,
+    sysroot: &Path,
+    target: &Target,
+    mount_prefix: &Path,
+    verbose: bool,
+) -> Result<()> {
+    // the rust toolchain is quite large, but most of it isn't needed
+    // we need the bin, libexec, and etc directories, and part of the lib directory.
+    let dst = mount_prefix.join("rust");
+    create_volume_dir(engine, container, &dst, verbose)?;
+    for basename in ["bin", "libexec", "etc"] {
+        let file = sysroot.join(basename);
+        copy_volume_files(engine, container, &file, &dst, verbose)?;
+    }
+
+    // the lib directories are rather large, so we want only a subset.
+    // now, we use a temp directory for everything else in the libdir
+    // we can pretty safely assume we don't have symlinks here.
+    let rustlib = Path::new("lib").join("rustlib");
+    let src_rustlib = sysroot.join(&rustlib);
+    let dst_rustlib = dst.join(&rustlib);
+
+    // SAFETY: safe, single-threaded execution.
+    let tempdir = unsafe { temp::TempDir::new()? };
+    let temppath = tempdir.path();
+    copy_dir(&sysroot.join("lib"), temppath, 0, |e, d| {
+        d == 0 && e.file_name() == "rustlib"
+    })?;
+    fs::create_dir(&temppath.join("rustlib")).ok();
+    copy_dir(
+        &src_rustlib,
+        &temppath.join("rustlib"),
+        0,
+        |entry, depth| {
+            if depth != 0 {
+                return false;
+            }
+            let file_type = match entry.file_type() {
+                Ok(file_type) => file_type,
+                Err(_) => return true,
+            };
+            let file_name = entry.file_name();
+            !(file_type.is_file() || file_name == "src" || file_name == "etc")
+        },
+    )?;
+    copy_volume_files(engine, container, temppath, &dst.join("lib"), verbose)?;
+    // must make the `dst.join("lib")` **after** here, or we copy temp into lib.
+    create_volume_dir(engine, container, &dst_rustlib, verbose)?;
+
+    // we first copy over the toolchain file, then everything besides it.
+    // since we don't want to call docker 100x, we copy the intermediate
+    // files to a temp directory so they're cleaned up afterwards.
+    let toolchain_path = src_rustlib.join(&target.triple());
+    if toolchain_path.exists() {
+        copy_volume_files(engine, container, &toolchain_path, &dst_rustlib, verbose)?;
+    }
+
+    // now we need to copy over the host toolchain too, since it has
+    // some requirements to find std libraries, etc.
+    let rustc = sysroot.join("bin").join("rustc");
+    let libdir = Command::new(rustc)
+        .args(&["--print", "target-libdir"])
+        .run_and_get_stdout(verbose)?;
```

Comment:
this assumes the host is `x86_64-unknown-linux-gnu`

on my mac, I'm getting 

````
emil@Emils-MacBook-Pro rust-cpp-hello-word % CROSS_REMOTE=1 cross -v build --target x86_64-unknown-linux-gnu
+ cargo metadata --format-version 1 --filter-platform x86_64-unknown-linux-gnu
+ rustc --print sysroot
+ rustup toolchain list
+ rustup target list --toolchain stable-x86_64-unknown-linux-gnu
+ rustup component list --toolchain stable-x86_64-unknown-linux-gnu
+ /usr/local/bin/docker
+ /usr/local/bin/docker volume inspect cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e-keep
+ /usr/local/bin/docker ps -a --filter 'name=cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e' --format {{.State}}
+ /usr/local/bin/docker volume inspect cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker volume create cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker run --userns host --name cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e -v cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross -e 'PKG_CONFIG_ALLOW_CROSS=1' -e 'XARGO_HOME=/xargo' -e 'CARGO_HOME=/cargo' -e 'CARGO_TARGET_DIR=/target' -e 'CROSS_RUNNER=' -e 'USER=emil' -v /cross/cargo/bin -d -t ghcr.io/cross-rs/x86_64-unknown-linux-gnu:main sh -c 'sleep infinity'
394cab661e2fc43808f47e32065124403ce3e58a83740565d489ad4ae62aacf6
+ /usr/local/bin/docker exec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e sh -c 'mkdir -p '\''/cross/cargo'\'''
+ /usr/local/bin/docker cp -a /Users/emil/.cargo/bin cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/cargo
+ /usr/local/bin/docker cp -a /Users/emil/.cargo/env cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/cargo
+ /usr/local/bin/docker exec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e sh -c 'mkdir -p '\''/cross/rust'\'''
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/libexec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/etc cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust
+ /usr/local/bin/docker cp -a '/Users/emil/Library/Application Support/cross-rs/tmp/.tmpZBMzqs' cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust/lib
+ /usr/local/bin/docker exec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e sh -c 'mkdir -p '\''/cross/rust/lib/rustlib'\'''
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust/lib/rustlib
+ /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --print target-libdir
+ /usr/local/bin/docker stop cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker rm cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker volume rm cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
Error: 
   0: `/Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --print target-libdir` failed with exit status: 126

Stderr:
   /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc: /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc: cannot execute binary file
```

Reply:
or, assumes the host can run the toolchain

Reply:
this assumes the host is `x86_64-unknown-linux-gnu`

on my mac, I'm getting 

````
emil@Emils-MacBook-Pro rust-cpp-hello-word % CROSS_REMOTE=1 cross -v build --target x86_64-unknown-linux-gnu
+ cargo metadata --format-version 1 --filter-platform x86_64-unknown-linux-gnu
+ rustc --print sysroot
+ rustup toolchain list
+ rustup target list --toolchain stable-x86_64-unknown-linux-gnu
+ rustup component list --toolchain stable-x86_64-unknown-linux-gnu
+ /usr/local/bin/docker
+ /usr/local/bin/docker volume inspect cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e-keep
+ /usr/local/bin/docker ps -a --filter 'name=cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e' --format {{.State}}
+ /usr/local/bin/docker volume inspect cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker volume create cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker run --userns host --name cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e -v cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross -e 'PKG_CONFIG_ALLOW_CROSS=1' -e 'XARGO_HOME=/xargo' -e 'CARGO_HOME=/cargo' -e 'CARGO_TARGET_DIR=/target' -e 'CROSS_RUNNER=' -e 'USER=emil' -v /cross/cargo/bin -d -t ghcr.io/cross-rs/x86_64-unknown-linux-gnu:main sh -c 'sleep infinity'
394cab661e2fc43808f47e32065124403ce3e58a83740565d489ad4ae62aacf6
+ /usr/local/bin/docker exec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e sh -c 'mkdir -p '\''/cross/cargo'\'''
+ /usr/local/bin/docker cp -a /Users/emil/.cargo/bin cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/cargo
+ /usr/local/bin/docker cp -a /Users/emil/.cargo/env cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/cargo
+ /usr/local/bin/docker exec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e sh -c 'mkdir -p '\''/cross/rust'\'''
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/libexec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/etc cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust
+ /usr/local/bin/docker cp -a '/Users/emil/Library/Application Support/cross-rs/tmp/.tmpZBMzqs' cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust/lib
+ /usr/local/bin/docker exec cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e sh -c 'mkdir -p '\''/cross/rust/lib/rustlib'\'''
+ /usr/local/bin/docker cp -a /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e:/cross/rust/lib/rustlib
+ /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --print target-libdir
+ /usr/local/bin/docker stop cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker rm cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
+ /usr/local/bin/docker volume rm cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
cross-hellopp-x86_64-unknown-linux-gnu-27448-47e2d-fe5b13d681f25ee6474be29d748c65adcd91f69e
Error: 
   0: `/Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc --print target-libdir` failed with exit status: 126

Stderr:
   /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc: /Users/emil/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc: cannot execute binary file
```

Reply:
```
let host_toolchain_path = sysroot.join("lib/rustlib").join(target.triple());
```

should just work, if the toolchain is malformed, there's something wrong

Reply:
There's a need also for more than just the target toolchain, otherwise there's errors in trying to find `rust-src`. It always needs the directory with the `bin` subdirectory, of which there is only one (and refers to the host toolchain), however. So it needs both the target and host toolchains, I believe, always. I'll do more testing on Windows (not WSL) and macOS.

Reply:
I think there's a terminology error here.

The `toolchain` is what you use to compile a target, say `stable-x86_64-unknown-linux-gnu`, this toolchain is compiled to run on a host that runs x86_64-unknown-linux-gnu. The binaries for the toolchain live under same directory as `$(rustup which rustc --toolchain <toolchain>)`, the parent of this directory is the sysroot (`rustc --print sysroot`), and the target libraries live under `$sysroot/lib/rustlib/<target>`.

You're already doing everything that is needed for host toolchain + host-target + target. The only thing wrong here is that you assume that the host can run the toolchain.

Reply:
on a related note, it shouldn't be called `host_toolchain_path`, but simply `toolchain_path`

Reply:
Why? The assumption cross does today (which #817 tries to fix) is that all images run on x86_64-unknown-linux-gnu. If we don't replace it, we could possibly mount the hosts (the one running `CROSS_REMOTE=1 cross ...` ) toolchain, which could be aarch64, powerpc or whatever.

Reply:
No this is the issue for my code specifically, is that we need the toolchain that has the bin directory, which is the one for the toolchain we're currently using. Say for example, if we're using `x86_64-pc-windows-msvc`, it doesn't need it to compile, but it does need the other subdirectories, and that's why it views it's hardcoding `x86_64-unknown-linux-gnu`.

Reply:
> Say for example, if we're using x86_64-pc-windows-msvc

Is this the target or the "toolchain"?

Reply:
The toolchain. The default toolchain. We need both the target and the default toolchain. So if I'm cross-compiling to `arm-unknown-linux-gnueabihf` on `x86_64-unknown-linux-gnu`, I need both, but on `x86_64-pc-windows-msvc`, I need `arm-unknown-linux-gnueabihf` and `x86_64-pc-windows-msvc` but should not need `x86_64-unknown-linux-gnu`. This is because that directory has `rust-src` or `rust-std` (depending on if building from source or not) that are needed, and aren't present elsewhere in the specific toolchains.

Reply:
Aha! That's wrong and where the misunderstanding is.

To compile for a target, we don't need its toolchain, you only need the components you require in `sysroot/lib/rustlib/<target>`. A toolchain has only one binary folder.

The toolchain we use in our images are always of sort `.rustup/toolchains/<channel>-<opt_date>-x86_64-unknown-linux-gnu`. We call this the `sysroot`. The current sysroot function makes it so that we pull the same toolchain "version" for x86_64-unknown-linux-gnu as the host.
We can't use the toolchain compiled for `x86_64-pc-windows-msvc` inside our images (yet).

The intended effect of replacing the host_triple in `<channel>-<opt_date>-<host_triple>` with `x86_64-unknown-linux-gnu` is so that we know it can run on the images. We also make sure to `rustup toolchain add --toolchain <channel>-<opt_date>-x86_64-unknown-linux-gnu`

The binaries for a toolchain lives in `sysroot/bin`, the binaries looks for the needed components in `sysroot/lib/rustlib/<target>`. You don't have multiple toolchains, you have multiple targets in that toolchain. see https://rust-lang.github.io/rustup/cross-compilation.html

The issue with the current changeset is that it pulls in whatever the host toolchain is, that will only ever be valid when you are on `x86_64-unknown-linux-gnu` yourself


So, if you're on cross-compiling to `arm-unknown-linux-gnueabihf` on `x86_64-unknown-linux-gnu`, you only need

`rustup toolchain add x86_64-unknown-linux-gnu` and `rustup target add arm-unknown-linux-gnueabihf --toolchain x86_64-unknown-linux-gnu`

Reply:
ofc, as you pointed out, some components live in the host target folder, so that should always follow. But you already do that here, the only wrong thing is that we may not be able to run the `<channel>-<opt_date>-x86_64-unknown-linux-gnu`s rustc.

The call to `rustc --print target-libdir` is redundant, it's not needed, we already know where it would live

Reply:
I guess on non-Linux systems we probably will also need something more.

Reply:
No, not really.